src/client/executor.rs

```diff
@@ -363,36 +443,112 @@ impl Client {
+                    {
+                        return Ok(Retryability::Write);
+                    }
+                }
+                _ => {}
             }
-            _ => Ok(Retryability::None),
         }
+        Ok(Retryability::None)
     }
 }
 
-/// Returns an Error with a "RetryableWriteError" label added if necessary. On a pre-4.4
-/// connection, a label should be added to any write-retryable error. On a 4.4+ connection, a
-/// label should only be added to network errors. Regardless of server version, a label should
-/// only be added if the `retry_writes` client option is not set to `false`.
-async fn get_error_with_retryable_write_label(conn: &Connection, err: Error) -> Result<Error> {
-    if let Some(max_wire_version) = conn.stream_description()?.max_wire_version {
-        if err.should_add_retryable_write_label(max_wire_version) {
-            return Ok(err.with_label("RetryableWriteError"));
+impl Error {
+    /// Adds the necessary labels to this Error.
+    ///
+    /// A TransientTransactionError label should be added if a transaction is in progress and the
+    /// error is a network or server selection error.
+    ///
+    /// On a pre-4.4 connection, a RetryableWriteError label should be added to any write-retryable
+    /// error. On a 4.4+ connection, a label should only be added to network errors. Regardless of
+    /// server version, a label should only be added if the `retry_writes` client option is not set
+    /// to `false`, the operation during which the error occured is write-retryable, and a
+    /// TransientTransactionError label has not already been added.
+    fn add_labels(
+        &mut self,
+        conn: Option<&Connection>,
+        session: &Option<&mut ClientSession>,
+        retryability: Option<&Retryability>,
+    ) -> Result<()> {
+        let transaction_state = session.as_ref().map_or(&TransactionState::None, |session| {
+            &session.transaction.state
+        });
+        match transaction_state {
+            TransactionState::Starting | TransactionState::InProgress => {
+                if self.is_network_error() || self.is_server_selection_error() {
+                    self.add_label(TRANSIENT_TRANSACTION_ERROR);
+                }
+            }
+            TransactionState::Committed { .. } => {
+                if let Some(conn) = conn {
+                    if let Some(max_wire_version) = conn.stream_description()?.max_wire_version {
+                        if self.should_add_retryable_write_label(max_wire_version) {
+                            self.add_label(RETRYABLE_WRITE_ERROR);
+                        }
+                    }
+                }
+                if self.should_add_unknown_transaction_commit_result_label() {
+                    self.add_label(UNKNOWN_TRANSACTION_COMMIT_RESULT);
+                }
+            }
+            TransactionState::Aborted => {
+                self.add_label(RETRYABLE_WRITE_ERROR);
```

Comment:
For abort, we don't need to check if the client supports retryability, but we do need to ensure the error does.
Per the spec:
>abortTransaction is a retryable write command. Drivers MUST retry once
after abortTransaction fails with a retryable error according to the
`Retryable Writes Specification`_., regardless of whether retryWrites is set
on the MongoClient or not.